File: Algorithms/0384_Shuffle_an_Array/Python/Shuffle_an_Array_Solution_1.py

# Space: O(n)
# Time: O(n)

import random


class Solution:

    def __init__(self, nums):
        self.original = nums

    def reset(self):
        """
        Resets the array to its original configuration and return it.
        """
        return self.original

    def shuffle(self):
        """
        Returns a random shuffling of the array.
        """
        res = []
        temp = self.original[:]
        while temp:
            temp_num = random.choice(temp)
            res.append(temp_num)
            temp.remove(temp_num)
        return res

File: Algorithms/0384_Shuffle_an_Array/Java/ShuffleanArraySolution1.java


// Space: O(1)
// Time: O(n)

public class Solution {
    private int[] original;

    public Solution(int[] nums) {
        original = nums;
    }

    /** Resets the array to its original configuration and return it. */
    public int[] reset() {
        return original;
    }

    /** Returns a random shuffling of the array. */
    public int[] shuffle() {
        int[] res = new int[original.length];
        List<Integer> cache = new ArrayList<>();
        Random rand = new Random();
        for (int i : original) {
            cache.add(i);
        }
        for (int i = 0; i < original.length; i++) {
            int tempIndex = rand.nextInt(cache.size());
            res[i] = cache.get(tempIndex);
            cache.remove(tempIndex);
        }
        return res;
    }
}




